# piRNA Differential Expression

```{r}
# load in the counts matrix generated by the script from the Le Greca et al paper's github
if (!file.exists('rds/pirna_counts.rds')) {
	piRNA_counts_original = read.table("Data/piRNA_counts/original_batch_LeGreca_piRNA.counts.gz", header = T, sep = "\t")
	piRNA_counts_new = read.table("Data/piRNA_counts/new_batch_LeGreca_piRNA.counts.gz", header = T, sep = "\t")
	pirna_counts = merge(piRNA_counts_original, piRNA_counts_new, by = "Geneid", all = TRUE)
	rm(piRNA_counts_original)
	rm(piRNA_counts_new)

	pirna_counts$Chr.x = NULL
	pirna_counts$Start.x = NULL
	pirna_counts$End.x = NULL
	pirna_counts$Strand.x = NULL
	pirna_counts$Length.x = NULL
	pirna_counts$Chr.y = NULL
	pirna_counts$Start.y = NULL
	pirna_counts$End.y = NULL
	pirna_counts$Strand.y = NULL
	pirna_counts$Length.y = NULL
	colnames(pirna_counts) = gsub("^([A-Z][1-9])_S.*", "\\1", colnames(pirna_counts), perl = T)
	colnames(pirna_counts) = gsub("^(SL[1-9]+_[A-Z])Aligned.sorted.*", "\\1", colnames(pirna_counts), perl = T)
	colnames(pirna_counts) = gsub("Aligned.sortedByCoord.out.SL290_GAligned.sortedByCoord.out.lenFilt.sam$","", colnames(pirna_counts))
	rownames(pirna_counts) = pirna_counts$Geneid
	pirna_counts$Geneid = NULL

	# remove rows with zero piRNAs
	pirna_counts$rowsums = rowSums(pirna_counts, na.rm = T)
	pirna_counts = pirna_counts[(pirna_counts$rowsums > 0),]
	pirna_counts$rowsums = NULL

	# remove piRNAs present in less than 2 samples
	pirna_counts[is.na(pirna_counts)] <- 0
	pirna_counts[(pirna_counts == 0)] <- NA
	pirna_counts$innbcells = rowSums(!is.na(pirna_counts))
	pirna_counts = pirna_counts[(pirna_counts$innbcells > 2),]
	pirna_counts$innbcells = NULL
	pirna_counts[is.na(pirna_counts)] <- 0

	pirna_counts = pirna_counts[grepl("piR-hsa-", rownames(pirna_counts)),]
	saveRDS(pirna_counts, file = 'rds/pirna_counts.rds')
} else {
	pirna_counts = readRDS('rds/pirna_counts.rds')
}
```


## PCA of All Annotations


```{r cache=TRUE, dev="png"}
submerged = testless_annot
rownames(submerged) = submerged$Row.names
# submerged$Row.names = NULL
submerged$Hist_Type_wt_ovary_distinction = as.character(submerged$Hist_Type)
submerged$Hist_Type_wt_ovary_distinction[submerged$tissue_type == "Testis" & submerged$Hist_Type == "Gonadal control"] = "Gonadal control (testis)"
submerged$Hist_Type_wt_ovary_distinction[submerged$tissue_type == "Ovary" & submerged$Hist_Type == "Gonadal control"] = "Gonadal control (ovary)"

equalised_tables = equalise_annot_counts(submerged, pirna_counts)
deseq_input_annot = equalised_tables[[1]]
local_trimmed_pirna_counts = equalised_tables[[2]]
local_trimmed_pirna_counts = local_trimmed_pirna_counts[,rownames(deseq_input_annot)]
local_trimmed_pirna_counts = local_trimmed_pirna_counts + 1

# load data as DeSeq
deseq_matrix_raw = DESeqDataSetFromMatrix(countData = local_trimmed_pirna_counts,
										  colData = deseq_input_annot,
										  design = ~ Hist_Type + batch)
rm(equalised_tables)
rm(deseq_input_annot)
rm(submerged)

vsd <- vst(deseq_matrix_raw, nsub = 500)

# Remove batch effect
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)
assay(vsd) <- mat
rm(mat)

# Plot PCA
set.seed(136)
pca_values = plotPCA(vsd, intgroup=c("tissue_type","Row.names"), returnData=T)

plt = ggplot(data = pca_values, aes_string(x = "PC1", y = "PC2", fill = "tissue_type", label = "Row.names")) +
  geom_point(size = 4, shape = 21, alpha=0.7) +
  xlab(paste0("PC1: ", round(attr(pca_values, "percentVar")[1] * 100), "% variance")) +
  ylab(paste0("PC2: ", round(attr(pca_values, "percentVar")[2] * 100), "% variance")) +
  geom_text_repel(size = 2, max.overlaps = 100) +
  coord_fixed() +
  ggtitle("PCA of both cohorts (sans test-lane) piRNA expression coloured by tissue_type, labelled by cell")

pirna_tissue_PCA = ggplot(data = pca_values, aes_string(x = "PC1", y = "PC2", fill = "tissue_type", label = "tissue_type")) +
  geom_point(size = 4, shape = 21, alpha=0.7) +
  xlab(paste0("PC1: ", round(attr(pca_values, "percentVar")[1] * 100), "% variance")) +
  ylab(paste0("PC2: ", round(attr(pca_values, "percentVar")[2] * 100), "% variance")) +
  geom_text_repel(size = 2, max.overlaps = 100) +
  coord_fixed() +
  ggtitle("PCA of both cohorts (sans test-lane) piRNA expression coloured by tissue_type")

```


```{r cache=TRUE, dev="png"}
# generate a PCA colored by each of the annotation columns
for (annotcol in colnames(colData(vsd))[!colnames(colData(vsd)) %in% c("Row.names")]) {
	plt = plotPCA(vsd, intgroup=c(annotcol))
	plt = plt + ggtitle(paste0("PCA of piRNA counts, colored by : ", annotcol))
	if (annotcol == "Hist_Type_wt_ovary_distinction") {
    set.seed(136)
    pca_values = plotPCA(vsd, intgroup=c("Hist_Type_wt_ovary_distinction","Type"), returnData=T)

    plt = ggplot(data = pca_values, aes_string(x = "PC1", y = "PC2", colour = "Hist_Type_wt_ovary_distinction", shape = "Type")) +
      geom_point(size = 4) +
      histological_subtype_palette_wt_ovary_distinction +
      xlab(paste0("PC1: ", round(attr(pca_values, "percentVar")[1] * 100), "% variance")) +
      ylab(paste0("PC2: ", round(attr(pca_values, "percentVar")[2] * 100), "% variance")) +
      coord_fixed()

    master_plot_list$figS9 = plt
	}

	print(plt)
	ggsave(filename = paste0("Plots/piRNA Clustering/PCA_",annotcol,".pdf"), device = "pdf", width = 8, height = 5)
}
```

```{r fig.width=6.22, fig.height=3.86}
# for poster presentation
pirna_tissue_PCA
ggsave(filename = "Plots/piRNA Clustering/piRNA_PCA.svg", device = "svg",width = 6.22, height = 3.86)
```

```{r}
# generate a UMAP colored by each of the annotation columns
for (annotcol in colnames(colData(vsd))[!colnames(colData(vsd)) %in% c("Row.names")]) {
  set.seed(126)
  # generate UMAP_1 and UMAP_2 from count data
  umap_coordinates = umap(t(assay(vsd)))
  umap_results = cbind(data.frame(umap_coordinates$layout), vsd[[annotcol]])
  colnames(umap_results) = c("UMAP_1", "UMAP_2", "Group")

  plt = ggplot(umap_results, aes(x = UMAP_1, y = UMAP_2, colour = Group)) +
    geom_point()

	plt = plt + ggtitle(paste0("UMAP_ of piRNA counts, colored by : ", annotcol))
	if (annotcol == "Hist_Type") {
	  plt = plt + histological_subtype_palette
	}
	ggsave(filename = paste0("Plots/piRNA Clustering/UMAP_",annotcol,".pdf"), device = "pdf", width = 8, height = 5)
}
```



## All PCs from previous DeSeq in multiplot

```{r}
pirna_cells_annot <- as.data.frame(colData(vsd))
pirna_cells_annot$Row.names = NULL
set.seed(136)
all_pcs <- prcomp(t(assay(vsd)), scale. = T)
pcs <- c(1, 2, 3, 4, 5, 6, 7, 8)
rm(vsd)
```


### All PCs for each Annot

```{r cache=TRUE, dev="png"}
dir.create("Plots/piRNA Clustering/21-04-All PCs per Annot", showWarnings = F, recursive = T)
for (annotcol in colnames(pirna_cells_annot)) {
  for (pc in pcs) {
    plt <- autoplot(all_pcs, x = pc, y = pc + 1, data = pirna_cells_annot, colour = annotcol)
    plt <- plt + ggtitle(paste0("PCA of piRNA counts, colored by : ", annotcol))
    eval(parse(text = paste("p", pc, " = plt", sep = ""))) # set plt to be p1,p2, etc based on i
  }
  list_of_plots <- ls(pattern = "^p[0-9]+") # get all pX objects we just created
  plot_pc_grid <- plot_grid(plotlist = mget(list_of_plots), labels = "AUTO") # the mget gets objects by names in list

  eval(parse(text = paste0("rm(", paste(list_of_plots, collapse = ", "), ")"))) # delete all objects in list_of_plots
  print(plot_pc_grid)
  ggsave(filename = paste0("Plots/piRNA Clustering/21-04-All PCs per Annot/PCA_", annotcol, ".pdf"), device = "pdf", width = 20, height = 9)
}
```


### All Annots for each PC

```{r cache=TRUE}
dir.create("Plots/piRNA Clustering/21-04-All Annots per PC", showWarnings = F, recursive = T)
for (pc in pcs) {
  i <- 1
  for (annotcol in colnames(pirna_cells_annot)) {
    plt <- autoplot(all_pcs, x = pc, y = pc + 1, data = pirna_cells_annot, colour = annotcol)
    plt <- plt + ggtitle(paste0("PCA of piRNA counts, colored by : ", annotcol))
    eval(parse(text = paste("p", i, " = plt", sep = ""))) # set plt to be p1,p2, etc based on i
    i <- i + 1
  }
  rm(plt)
  list_of_plots <- ls(pattern = "^p[0-9]+") # get all pX objects we just created
  plot_pc_grid <- plot_grid(plotlist = mget(list_of_plots), labels = "AUTO") # the mget gets objects by names in list

  eval(parse(text = paste0("rm(", paste(list_of_plots, collapse = ", "), ")"))) # delete all objects in list_of_plots
  print(plot_pc_grid)
  ggsave(filename = paste0("Plots/piRNA Clustering/21-04-All Annots per PC/PCA_for_PC", pc, "_PC", pc + 1, ".pdf"), device = "pdf", width = 25, height = 11)
}
```


## Test/Final lane Comparison {-}

```{r cache=TRUE}
# load data as DeSeq
original_annot = annot_table[annot_table$batch == "original",]
rownames(original_annot) = original_annot$Row.names


equalised_tables = equalise_annot_counts(original_annot, pirna_counts)
deseq_input_annot = equalised_tables[[1]]
local_trimmed_pirna_counts = equalised_tables[[2]]
local_trimmed_pirna_counts = local_trimmed_pirna_counts[,rownames(deseq_input_annot)]
local_trimmed_pirna_counts = local_trimmed_pirna_counts + 1



deseq_matrix_raw = suppressMessages(DESeqDataSetFromMatrix(countData = local_trimmed_pirna_counts,
										  colData = deseq_input_annot,
										  design = ~ Run_number))

vsd <- vst(deseq_matrix_raw, nsub=10)

# Plot PCA
set.seed(136)
plt = plotPCA(vsd, intgroup=c("Run_number")) + ggtitle("PCA of cohort-1 piRNA coloured by lane")
ggsave(filename = "Plots/piRNA Clustering/PCA_lane_separation.pdf", device = "pdf", width = 8, height = 5)
rm(original_annot)
rm(original_pirna)
plt
```


## Batch Comparison {-}

```{r cache=TRUE}
submerged = testless_annot
rownames(submerged) = submerged$Row.names

# load data as DeSeq
equalised_tables = equalise_annot_counts(submerged, pirna_counts)
deseq_input_annot = equalised_tables[[1]]
local_trimmed_pirna_counts = equalised_tables[[2]]
local_trimmed_pirna_counts = local_trimmed_pirna_counts[,rownames(deseq_input_annot)]
local_trimmed_pirna_counts = local_trimmed_pirna_counts + 1

deseq_matrix_raw = suppressMessages(DESeqDataSetFromMatrix(countData = local_trimmed_pirna_counts,
										  colData = deseq_input_annot,
										  design = ~ Hist_Type + batch))

vsd <- vst(deseq_matrix_raw, nsub=500)

# Remove batch effect
pirna_norm_counts <- assay(vsd)
pirna_norm_counts <- limma::removeBatchEffect(pirna_norm_counts, vsd$batch)
assay(vsd) <- pirna_norm_counts

# Plot PCA
set.seed(136)
plt = plotPCA(vsd, intgroup=c("batch")) + ggtitle("PCA of piRNA counts (both cohorts) coloured by batch")
ggsave(filename = "Plots/piRNA Clustering/PCA_batch_separation.pdf", device = "pdf", width = 8, height = 5)
rm(local_trimmed_pirna_counts)
plt
```

## PCA of just Gonadol controls

```{r cache=TRUE}
submerged = testless_annot
rownames(submerged) = submerged$Row.names
submerged$Row.names = NULL

submerged = submerged[submerged$Hist_Type == "Gonadal control",]


equalised_tables = equalise_annot_counts(submerged, pirna_counts)
deseq_input_annot = equalised_tables[[1]]
local_trimmed_pirna_counts = equalised_tables[[2]]
local_trimmed_pirna_counts = local_trimmed_pirna_counts[,rownames(deseq_input_annot)]
local_trimmed_pirna_counts = local_trimmed_pirna_counts + 1

# load data as DeSeq
deseq_matrix_raw = DESeqDataSetFromMatrix(countData = local_trimmed_pirna_counts,
										  colData = deseq_input_annot,
										  design = ~ tissue_type + batch)
rm(equalised_tables)
rm(local_trimmed_pirna_counts)
rm(deseq_input_annot)
rm(submerged)

vsd <- vst(deseq_matrix_raw, nsub = 500)

# Remove batch effect
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)
assay(vsd) <- mat
rm(mat)
```



```{r cache=TRUE}
# Plot PCA
set.seed(136)
plt1 = plotPCA(vsd, intgroup=c("tissue_type"), )
plt1$layers[[1]]$aes_params$alpha = 0.25 # make alpha of all points 0.25
plt1 = plt1 + geom_text_repel(aes(label=colnames(vsd)))
plt1 = plt1 + ggtitle(paste0("PCA of piRNA counts in only gonadol controls, colored by : tissue_type"))
plt2 = plotPCA(vsd, intgroup=c("Tumour"), )
plt2 = plt2 + ggtitle(paste0("PCA of piRNA counts in only gonadol controls, colored by : Tumour"))
plt = plot_grid(plt1, plt2, ncol = 2)
rm(plt1)
rm(plt2)
ggsave(plot = plt, filename = paste0("Plots/piRNA Clustering/PCA_","only_gonadal_controls_TissueTumour",".pdf"), device = "pdf", width = 397, height = 92, units = "mm")
print(plt)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
# cleanup left behind objects in memory from our PCA generation
rm(list = c("plt","plot_pc_grid","pc","pcs","annotcol","pirna_cells_annot","all_pcs","vsd","i","list_of_plots"))
gc() # run garbage collection, clear ram after having deleted these objects
```




## Diff Expression Analysis

```{r}
pirna_i <- 0
```


### Ovary Normal / Testes Normal {.panelset}

```{r cache=TRUE}
grpA_label = "Ovary Normal"
Aint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Ovary" & annot_table$Hist_Type == "Gonadal control" & annot_table$Row.names %in% testless_cells),]))
grpB_label = "Testis Normal"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Testis" & annot_table$Hist_Type == "Gonadal control" & annot_table$Row.names %in% testless_cells),]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```


### Testes Tumor / Testes Normal {.panelset}

```{r cache=TRUE}
grpA_label = "Testis Tumour"
Aint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Testis" & annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells),]))
grpB_label = "Testis Normal"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Testis" & annot_table$Normal_or_Tumour == "Normal" & annot_table$Row.names %in% testless_cells),]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```



### Testes Primary Tumor / Lung Metastasis {.panelset}

```{r cache=TRUE}
grpA_label = "Testis Primary Tumour"
Aint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Testis" & annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells),]))
grpB_label = "Lung Metastasis"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Metastasis(lung)" & annot_table$Row.names %in% testless_cells),]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, fitType="mean", subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```



### Ovarian Primary Tumor / Lymph Node Metastasis {.panelset}


```{r cache=TRUE}
grpA_label = "Ovarian Primary Tumour"
Aint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Ovary" & annot_table$Normal_or_Tumour == "Tumour"  & annot_table$Row.names %in% testless_cells),]))
grpB_label = "Lymph Node Metastasis"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Metastasis(LN)"  & annot_table$Row.names %in% testless_cells),]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, fitType="mean", subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```




## Sean's piRNA Comparisons

### Tissue Healthy/Tumour {.panelset}

```{r cache=TRUE}
grpA_label = "Tumour tissue samples"
Aint = as.integer(rownames(annot_table[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour == "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane"),]))
Aint = c(Aint, as.integer(rownames(annot_table[(annot_table$tissue_type == "Metastasis(lung)" & annot_table$Run_number == "final-lane"),])))

grpB_label = "Healthy tissue samples"
Bint = as.integer(rownames(annot_table[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour != "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane"),]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```


### Tumour/CL {.panelset}

```{r cache=TRUE}
grpA_label = "Tumour samples and cell lines (wt EVs)"
Aint = as.integer(rownames(annot_table[annot_table$Type == "Cell line" & annot_table$Normal_or_Tumour == "Tumour",]))
Aint = c(Aint, as.integer(rownames(annot_table[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour == "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane"),])))
Aint = c(Aint, as.integer(rownames(annot_table[(annot_table$tissue_type == "Metastasis(lung)" & annot_table$Run_number == "final-lane"),])))

grpB_label = "Healthy samples and cell lines (wt EVs)"
Bint = as.integer(rownames(annot_table[annot_table$Type == "Cell line" & annot_table$Normal_or_Tumour == "Normal",]))
Bint = c(Bint, as.integer(rownames(annot_table[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour != "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane"),])))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

```{r fig.width=7.22, fig.height=4.86}
# for poster presentation
unserialize(deseq_out$volcano)
ggsave(filename = "Plots/piRNA DEA/piRNA_DE_volcano.svg", device = "svg",width = 7.22, height = 4.86)
```


#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```


### CL {.panelset}

```{r cache=TRUE}
grpA_label = "Tumour Cell lines (wt EVs)"
Aint = as.integer(rownames(annot_table[annot_table$Type == "Cell line" & annot_table$Normal_or_Tumour == "Tumour",]))
grpB_label = "Healthy Cell lines (wt EVs)"
Bint = as.integer(rownames(annot_table[annot_table$Type == "Cell line" & annot_table$Normal_or_Tumour == "Normal",]))

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```



## Site specific differences

### Testis YST tissue v Ovary YST tissue

two testicular YST cases (n=4 samples using the test and final lane technical replicates) and the two ovarian YST cases (again, n=4 samples using the test and final lane technical replicates). WE can then see whether there is a site difference. This is not possible by the other subtypes as they are either ovarian (SEM/DYSGERM) or testicular (EC).


```{r}
testis_v_ovary_annot = annot_table[annot_table$Tumour == "YST" & annot_table$tissue_type %in% c("Testis","Ovary"),]

grpA_label = "Testis tissue YST"
Aint = as.integer(rownames(testis_v_ovary_annot)[testis_v_ovary_annot$tissue_type == "Testis"])
grpB_label = "Ovary tissue YST"
Bint = as.integer(rownames(testis_v_ovary_annot)[testis_v_ovary_annot$tissue_type == "Ovary"])

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA", fitType = "mean")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```

```{r}
yst_ovary_testis = read.table("Plots/piRNA DEA/8. DE results table (unfiltered) Ovary tissue YST v Testis tissue YST.tsv", header = TRUE, sep = "\t")
datatable(yst_ovary_testis)
```

poor results from the previous comparison may be a lack of site-specific signatures, but may just be because the first batch doesn't give us the necessary power. check if its just the poor first batch by making two additional comparisons : Ovary YST tissue samples v Ovary control samples, Testis YST tissue samples v Testis control samples.

### Ovarian YST v Tissue Controls

```{r}
ovarian_yst_control_annot = testless_annot[testless_annot$Tumour == "YST" & testless_annot$tissue_type %in% c("Ovary"),]
ovarian_yst_control_annot = rbind(ovarian_yst_control_annot, testless_annot[testless_annot$malignantVs.normal == "Non-malignant" & testless_annot$tissue_type %in% c("Ovary"),])

grpA_label = "Ovary tissue YST"
Aint = as.integer(rownames(ovarian_yst_control_annot)[ovarian_yst_control_annot$malignantVs.normal == "Malignant"])
grpB_label = "Ovary tissue control"
Bint = as.integer(rownames(ovarian_yst_control_annot)[ovarian_yst_control_annot$malignantVs.normal == "Non-malignant"])

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

##### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

##### Table {-}

```{r}
datatable(read_excel(path = deseq_out$excel))
```

##### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```


### Testicular YST v Tissue Controls

```{r}
testis_yst_control_annot = testless_annot[testless_annot$Tumour == "YST" & testless_annot$tissue_type %in% c("Testis"),]
testis_yst_control_annot = rbind(testis_yst_control_annot, testless_annot[testless_annot$malignantVs.normal == "Non-malignant" & testless_annot$tissue_type %in% c("Testis"),])

grpA_label = "Testis tissue YST"
Aint = as.integer(rownames(testis_yst_control_annot)[testis_yst_control_annot$malignantVs.normal == "Malignant"])
grpB_label = "Testis tissue control"
Bint = as.integer(rownames(testis_yst_control_annot)[testis_yst_control_annot$malignantVs.normal == "Non-malignant"])

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA")
pirna_i = deseq_out$counter
```

##### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

##### Table {-}

```{r}
datatable(read_excel(path = deseq_out$excel))
```

##### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```



## Additional Requests 20/04


```{r}
dir.create("Plots/piRNA DEA/22:04 - Pairwise GCT Comparisons", showWarnings = F, recursive = T)
main_pirna_i = pirna_i # backup piRNA_i counter as we will continue on from this for the Cell Line comparisons
pirna_i = 0
```


### All Testicular GCT against all healthy testis {.panelset}

```{r cache=TRUE}
grpA_label = "All Testis Tumour (inc lung metastasis)"
Aint = as.integer(rownames(annot_table[((annot_table$tissue_type == "Testis" | annot_table$tissue_type == "Metastasis(lung)") & (annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells)),]))
testes_tumour_ids = Aint
grpB_label = "All Testis Normal"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Testis" & annot_table$Normal_or_Tumour == "Normal"& annot_table$Row.names %in% testless_cells),]))
testes_control_ids = Bint

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA/22:04 - Pairwise GCT Comparisons")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```

#### Heatmap
for the manuscript
```{r}
tgct_v_t = read_excel(path = deseq_out$excel)
tgct_v_t = tgct_v_t[!is.na(tgct_v_t$padj),]
tgct_v_t = tgct_v_t[tgct_v_t$padj<0.05,]
tgct_v_t = tgct_v_t[abs(tgct_v_t$log2FoldChange) > 1,]
```

```{r}
deseq_matrix_raw = DESeqDataSetFromMatrix(countData = (pirna_counts[,colnames(pirna_counts) %in% annot_table$Row.names[c(Aint,Bint)]] + 1),
										  colData = annot_table[c(Aint,Bint),],
										  design = ~ Hist_Type)

deseq_matrix <- DESeq(deseq_matrix_raw)

vsd <- vst(deseq_matrix_raw, nsub=500)

# Remove batch effect
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)
assay(vsd) <- mat
rm(mat)

heatmap_annot = annot_table[c(Aint,Bint),]
rownames(heatmap_annot) = heatmap_annot$Row.names
heatmap_annot = heatmap_annot[,c("Hist_Type","tissue_type")]
colnames(heatmap_annot)[2] = "tissue_type"

Hist_Type_values=c("Yolk sac tumour" = "#FEC539",
		  		 "Seminoma/Germinoma" = "#008FFF",
		  		 "Embryonal carcinoma" = "red",
		  		 "Teratoma" = "brown",
		  		 "Gonadal control" = "#00A34E",
		  		 "Testis fibroblasts" = "grey"
		  		 )
heatmapcolors <- list(Hist_Type = Hist_Type_values)
rm(Hist_Type_values)


pheatmap(assay(vsd)[tgct_v_t$...1,],
          cluster_rows = FALSE,
          show_rownames = T,
          cluster_cols = T,
          show_colnames = T,
          annotation_col = heatmap_annot,
          border_color=NA,
          annotation_colors = heatmapcolors,
           main = "Differentially Expressed piRNAs between Malignant/Control tissue samples",
         )
```



### All Ovarian GCT against all healthy ovarian {.panelset}

```{r cache=TRUE}
grpA_label = "All ovarian tumour (inc lymph node metastasis)"
Aint = as.integer(rownames(annot_table[((annot_table$tissue_type == "Ovary" | annot_table$tissue_type == "Metastasis(LN)") & (annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells)),]))
ovary_tumour_ids = Aint
grpB_label = "All ovarian Normal"
Bint = as.integer(rownames(annot_table[(annot_table$tissue_type == "Ovary" & annot_table$Normal_or_Tumour == "Normal"& annot_table$Row.names %in% testless_cells),]))
ovary_control_ids = Bint

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA/22:04 - Pairwise GCT Comparisons")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```

#### Heatmap
for the manuscript
```{r}
ogct_v_o = read_excel(path = deseq_out$excel)
ogct_v_o = ogct_v_o[!is.na(ogct_v_o$padj),]
ogct_v_o = ogct_v_o[ogct_v_o$padj<0.05,]
ogct_v_o = ogct_v_o[abs(ogct_v_o$log2FoldChange) > 1,]
```

```{r eval=FALSE, include=FALSE}
deseq_matrix_raw = DESeqDataSetFromMatrix(countData = (pirna_counts[,colnames(pirna_counts) %in% annot_table$Row.names[c(Aint,Bint)]] + 1),
										  colData = annot_table[c(Aint,Bint),],
										  design = ~ Hist_Type)

deseq_matrix <- DESeq(deseq_matrix_raw)

vsd <- vst(deseq_matrix_raw, nsub=100)

# Remove batch effect
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)
assay(vsd) <- mat
rm(mat)

heatmap_annot = annot_table[c(Aint,Bint),]
rownames(heatmap_annot) = heatmap_annot$Row.names
heatmap_annot = heatmap_annot[,c("Hist_Type","tissue_type")]
colnames(heatmap_annot)[2] = "tissue_type"

Hist_Type_values=c("Yolk sac tumour" = "#FEC539",
		  		 "Seminoma/Germinoma" = "#008FFF",
		  		 "Embryonal carcinoma" = "red",
		  		 "Teratoma" = "brown",
		  		 "Gonadal control" = "#00A34E",
		  		 "Testis fibroblasts" = "grey"
		  		 )
heatmapcolors <- list(Hist_Type = Hist_Type_values)
rm(Hist_Type_values)


pheatmap(assay(vsd)[ogct_v_o$...1,],
          cluster_rows = FALSE,
          show_rownames = T,
          cluster_cols = T,
          show_colnames = T,
          annotation_col = heatmap_annot,
          border_color=NA,
          annotation_colors = heatmapcolors,
           main = "Differentially Expressed piRNAs between Malignant/Control tissue samples",
         )
```




### All Testicular GCT against EACH healthy testis

for these comparing against singular cells we have to use fitType="mean" instead of the default "parametric" model for dispersion estimation which is recommended for RNAseq

```{r cache=TRUE}
grpA_label = "All testis tumour (inc lung metastasis)"
Aint = as.integer(rownames(annot_table[((annot_table$tissue_type == "Testis" | annot_table$tissue_type == "Metastasis(lung)") & (annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells)),]))

for (i in 1:length(testes_control_ids)) {
	grpB_label = paste("Testis Normal, Cell", annot_table$Row.names[testes_control_ids[i]])
	Bint = testes_control_ids[i]
	deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts,forcenobatch=T, fitType="mean",  subfolder = "piRNA DEA/22:04 - Pairwise GCT Comparisons/3a - All TGCT v each healthy testis")
  pirna_i = deseq_out$counter
  unserialize(deseq_out$comptable)
  unserialize(deseq_out$volcano)
}
```


### All Ovarian GCT against EACH healthy ovarian

```{r cache=TRUE}
grpA_label = "All ovarian tumour (inc lymph node metastasis)"
Aint = as.integer(rownames(annot_table[((annot_table$tissue_type == "Ovary" | annot_table$tissue_type == "Metastasis(LN)") & (annot_table$Normal_or_Tumour == "Tumour" & annot_table$Row.names %in% testless_cells)),]))

for (i in 1:length(ovary_control_ids)) {
	grpB_label = paste("Ovarian Normal, Cell", annot_table$Row.names[ovary_control_ids[i]])
	Bint = ovary_control_ids[i]
	deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts,forcenobatch=T, fitType="mean",  subfolder = "piRNA DEA/22:04 - Pairwise GCT Comparisons/3c - All OGCT v each healthy ovarian")
  pirna_i = deseq_out$counter
  unserialize(deseq_out$comptable)
  unserialize(deseq_out$volcano)
}
```


## Ovary GCT v Testis GCT

```{r cache=TRUE}
grpA_label = "All Testis GCT"
Aint = annot_table %>% filter(Row.names %in% testless_cells) %>% filter(tissue_type %in% c("Testis","Metastasis(lung)")) %>% filter(malignantVs.normal=="Malignant") %>% pull(Row.names)
Aint = as.integer(rownames(annot_table[annot_table$Row.names %in% Aint,]))


grpB_label = "All Ovary GCT"
Bint = annot_table %>% filter(Row.names %in% testless_cells) %>% filter(tissue_type %in% c("Ovary","Metastasis(LN)")) %>% filter(malignantVs.normal=="Malignant") %>% pull(Row.names)
Bint = as.integer(rownames(annot_table[annot_table$Row.names %in% Bint,]))
testes_control_ids = Bint

deseq_out = deseq2_plot("piRNA", grpA_label, Aint, grpB_label, Bint, pirna_i, annot_table, pirna_counts, subfolder = "piRNA DEA/22:04 - Pairwise GCT Comparisons")
pirna_i = deseq_out$counter
```

#### Volcano {-}

```{r}
unserialize(deseq_out$volcano)
```

#### Table {-}

```{r}
#datatable(read_excel(path = deseq_out$excel))
```

#### Comparisons {-}

```{r}
unserialize(deseq_out$comptable)
```


## Heatmap of top piRNAs by DE

```{r cache=TRUE}
deseq_input_annot = annot_table
rownames(deseq_input_annot) = deseq_input_annot$Row.names
deseq_input_annot$Row.names = NULL

equalised_tables = equalise_annot_counts(deseq_input_annot, pirna_counts)
deseq_input_annot = equalised_tables[[1]]
trimmed_pirna_counts = equalised_tables[[2]]
trimmed_pirna_counts = trimmed_pirna_counts[,rownames(deseq_input_annot)]
trimmed_pirna_counts = trimmed_pirna_counts + 1


deseq_matrix_raw = DESeqDataSetFromMatrix(countData = trimmed_pirna_counts,
										  colData = deseq_input_annot,
										  design = ~ tissue_type + batch)
rm(equalised_tables)
rm(trimmed_pirna_counts)
rm(submerged)

vsd <- vst(deseq_matrix_raw, nsub = 100)

# Remove batch effect
original_cells = rownames(deseq_input_annot)[deseq_input_annot$batch == "original"]
new_cells = rownames(deseq_input_annot)[deseq_input_annot$batch == "new"]

mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)

orig_mat <- mat[,original_cells]
new_mat <- mat[,new_cells]
rm(vsd)

deseq_input_annot$malignantVs.normal = NULL # remove the annot cols we are not interested in
deseq_input_annot$Type = NULL
deseq_input_annot$Run_number = NULL
deseq_input_annot$Tumour = NULL
```



```{r}
output_folder = "Plots/piRNA DEA/Split piRNA Heatmaps/"
dir.create(output_folder, recursive = TRUE, showWarnings = FALSE)
annot_col_for_collabel = "Hist_Type"
```

```{r}
plotHeatmapsOfSelectedPiRNAs <- function(sortby_category, sample_origin, full_pirna_list, output_folder) {
  output_folder <- paste0(output_folder, "/", sample_origin, "/")
  dir.create(output_folder, showWarnings = F, recursive = T)

  # remove test-lane cells from original_cells list
  original_cells = annot_table$Row.names[annot_table$Row.names %in% original_cells & annot_table$Run_number == "final-lane"]
  orig_mat = orig_mat[,colnames(orig_mat) %in% original_cells]

  pirna_list <- full_pirna_list[1:100]
  pdf(paste0(output_folder, "Heatmap_top_", length(pirna_list), "_DE_piR_by_", sortby_category, "_in_", sample_origin, ".pdf"),
    width = 20, height = 20,
    title = paste0("Top ", length(pirna_list), " Differentially Expressed piRNAs")
  )
  print(Heatmap(orig_mat[pirna_list, ],
    column_labels = deseq_input_annot$Hist_Type[rownames(deseq_input_annot) %in% original_cells],
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 9),
    top_annotation = HeatmapAnnotation(df = deseq_input_annot[rownames(deseq_input_annot) %in% original_cells,], name = annot_col_for_collabel, col = cmplxhmapcols)
  ))
  print(Heatmap(new_mat[pirna_list, ],
    column_labels = deseq_input_annot$Hist_Type[rownames(deseq_input_annot) %in% new_cells],
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 9),
    top_annotation = HeatmapAnnotation(df = deseq_input_annot[rownames(deseq_input_annot) %in% new_cells,], name = annot_col_for_collabel, col = cmplxhmapcols)
  ))
  dev.off()

  print(Heatmap(orig_mat[pirna_list, ],
    column_labels = deseq_input_annot$Hist_Type[rownames(deseq_input_annot) %in% original_cells],
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 9),
    top_annotation = HeatmapAnnotation(df = deseq_input_annot[rownames(deseq_input_annot) %in% original_cells,], name = annot_col_for_collabel, col = cmplxhmapcols)
  ))
  print(Heatmap(new_mat[pirna_list, ],
    column_labels = deseq_input_annot$Hist_Type[rownames(deseq_input_annot) %in% new_cells],
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 9),
    top_annotation = HeatmapAnnotation(df = deseq_input_annot[rownames(deseq_input_annot) %in% new_cells,], name = annot_col_for_collabel, col = cmplxhmapcols)
  ))
}
```



### Tissues/Cell Lines (wt EVs)

```{r cache=TRUE}
table = read_excel("Plots/piRNA DEA/6. DE results table Healthy samples and cell lines (wt EVs) v Tumour samples and cell lines (wt EVs).xlsx")
table$absLFC = abs(table$log2FoldChange)
table = table[order(table$absLFC, decreasing = T),]
lfc_pirna_list = table$...1
table = table[order(table$padj, decreasing = F),]
padj_pirna_list = table$...1

plotHeatmapsOfSelectedPiRNAs(sortby_category = "log2FC", sample_origin = "tissues_wt_celllines", full_pirna_list = lfc_pirna_list, output_folder)
```

## piRNA Counts/Proportions

### piRNA Counts

```{r}
tumour_group = annot_table$Row.names[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour == "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane")]
tumour_group = c(tumour_group, annot_table$Row.names[(annot_table$tissue_type == "Metastasis(lung)" & annot_table$Run_number == "final-lane")])

healthy_group = annot_table$Row.names[annot_table$tissue_type== "Testis" & annot_table$Type == "Tissue" & annot_table$Normal_or_Tumour != "Tumour" & (is.na(annot_table$Run_number) | annot_table$Run_number == "final-lane")]

table = read_excel("Plots/piRNA DEA/6. DE results table Healthy samples and cell lines (wt EVs) v Tumour samples and cell lines (wt EVs).xlsx")
table$absLFC = abs(table$log2FoldChange)
table = table[table$absLFC > 1,]
table = table[table$padj < 0.05,]
table = table[order(table$absLFC, decreasing = T),]
full_lfc_pirna_list = table$...1 # full_lfc_pirna_list is all DE piRNAs between Healthy and Tumour
upreg_lfc_list = table$...1[table$log2FoldChange > 0]
downreg_lfc_list = table$...1[table$log2FoldChange < 0]
lfc_pirna_list = full_lfc_pirna_list[1:20]
```

```{r cache=TRUE}
pirna_counts$rownames = rownames(pirna_counts)
melted = reshape2::melt(pirna_counts[,c(tumour_group, healthy_group, "rownames")], id="rownames")
pirna_counts$rownames = NULL

melted$Group[melted$variable %in% healthy_group] = "Healthy"
melted$Group[melted$variable %in% tumour_group] = "Tumour"

melted$piR = "Other piRNAs"
melted$piR[melted$rownames %in% full_lfc_pirna_list] = "DE piRNAs"
melted$piR[melted$rownames %in% lfc_pirna_list] = "Top 20 DE piRNAs"
# melted$piR[melted$rownames %in% upreg_lfc_list] = "Upreg in Tumour"
# melted$piR[melted$rownames %in% downreg_lfc_list] = "Downreg in Tumour"
melted$piR = as.factor(melted$piR)
```



```{r cache=TRUE}
ggplot(melted[melted$rownames %in% upreg_lfc_list, ], aes(y = log2(value + 1), x = variable)) +
  facet_wrap(scales = "free", facets = ~Group) +
  geom_jitter(aes(color = piR, alpha = 0.05)) +
  geom_boxplot() +
  ggtitle("Read counts of piRNAs upreglated in tumour") +
  ylab("Read count in sample (log2 scale)") +
  xlab("Sample")
```

```{r cache=TRUE}
plotHeatmapsOfSelectedPiRNAs(sortby_category = "log2FC", sample_origin = "tissues_wt_celllines", full_pirna_list = upreg_lfc_list, output_folder)
```


```{r cache=TRUE}
ggplot(melted[melted$rownames %in% full_lfc_pirna_list, ], aes(y = log2(value + 1), x = variable)) +
  facet_wrap(scales = "free", facets = ~Group) +
  geom_jitter(aes(color = piR, alpha = 0.05)) +
  geom_boxplot() +
  ggtitle("Read counts of all DE piRNA in each sample") +
  ylab("Read count in sample (log2 scale)") +
  xlab("Sample")
```

```{r}
ggsave(filename = "Plots/piRNA DEA/Raw_counts_DE_piRNA.pdf", device = "pdf", width = 14, height = 7)
```




### piRNAs Proportions

```{r cache=TRUE}
melted2 = melted
melted2$value[melted2$variable == "SL286_E"] = (melted2$value[melted2$variable == "SL286_E"] / sum(pirna_counts[,"SL286_E"]))
melted2$value[melted2$variable == "SL287_A"] = (melted2$value[melted2$variable == "SL287_A"] / sum(pirna_counts[,"SL287_A"]))
melted2$value[melted2$variable == "SL289_C"] = (melted2$value[melted2$variable == "SL289_C"] / sum(pirna_counts[,"SL289_C"]))
melted2$value[melted2$variable == "B1"] = (melted2$value[melted2$variable == "B1"] / sum(pirna_counts[,"B1"]))
melted2$value[melted2$variable == "B2"] = (melted2$value[melted2$variable == "B2"] / sum(pirna_counts[,"B2"]))
melted2$value[melted2$variable == "B3"] = (melted2$value[melted2$variable == "B3"] / sum(pirna_counts[,"B3"]))
melted2$value[melted2$variable == "SL290_G"] = (melted2$value[melted2$variable == "SL290_G"] / sum(pirna_counts[,"SL290_G"]))
```

```{r cache=TRUE}
ggplot(melted2[melted2$rownames %in% full_lfc_pirna_list, ], aes(y = log2(value + 1), x = variable)) +
  facet_wrap(scales = "free", facets = ~Group) +
  geom_jitter(aes(color = piR, alpha = 0.05)) +
  geom_boxplot() +
  ggtitle("Proportion of Read counts of Top 20 DE piRNA") +
  ylab("Read count / total read count of sample") +
  xlab("Sample")
```

```{r cache=TRUE}
ggplot(melted2[melted2$rownames %in% upreg_lfc_list, ], aes(y = log2(value + 1), x = variable)) +
  facet_wrap(scales = "free", facets = ~Group) +
  geom_jitter(aes(color = piR, alpha = 0.05)) +
  geom_boxplot() +
  ggtitle("Proportion of Read counts of Top 20 DE piRNA") +
  ylab("Read count / total read count of sample") +
  xlab("Sample")
```

```{r}
ggsave(filename = "Plots/piRNA DEA/Proportion_counts_DE_piRNA.pdf", device = "pdf", width = 14, height = 7)
```
